File: spring_java/src/test/java/dev/pollito/spring_java/config/enums/EnumUtilsTest.java
package dev.pollito.spring_java.config.enums;

import static dev.pollito.spring_java.config.enums.EnumUtils.fromValue;
import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class EnumUtilsTest {

  private enum Color implements ValuedEnum<String> {
    RED("red"),
    GREEN("green"),
    BLUE("blue");

    private final String value;

    Color(String value) {
      this.value = value;
    }

    @Override
    public String getValue() {
      return value;
    }
  }

  @Test
  void fromValue_returnsMatchingConstant() {
    assertEquals(Color.RED, fromValue(Color.class, "red"));
    assertEquals(Color.GREEN, fromValue(Color.class, "green"));
    assertEquals(Color.BLUE, fromValue(Color.class, "blue"));
  }

  @Test
  void fromValue_throwsIllegalArgumentExceptionForUnknownValue() {
    IllegalArgumentException ex =
        assertThrows(IllegalArgumentException.class, () -> fromValue(Color.class, "yellow"));
    assertTrue(ex.getMessage().contains("Color"));
    assertTrue(ex.getMessage().contains("yellow"));
  }
}

File: spring_kotlin/src/test/kotlin/dev/pollito/spring_kotlin/config/enums/EnumUtilsTest.kt
package dev.pollito.spring_kotlin.config.enums

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class EnumUtilsTest {

  private enum class Color(private val code: String) : ValuedEnum<String> {
    RED("red"),
    GREEN("green"),
    BLUE("blue");

    override fun getValue(): String = code
  }

  private enum class Priority(private val level: Int) : ValuedEnum<Int> {
    LOW(1),
    MEDIUM(2),
    HIGH(3);

    override fun getValue(): Int = level
  }

  @Test
  fun `fromValue returns correct enum constant for matching string value`() {
    val result = EnumUtils.fromValue(Color::class.java, "green")
    assertEquals(Color.GREEN, result)
  }

  @Test
  fun `fromValue returns correct enum constant for matching int value`() {
    val result = EnumUtils.fromValue(Priority::class.java, 3)
    assertEquals(Priority.HIGH, result)
  }

  @Test
  fun `fromValue returns first constant when value matches first entry`() {
    val result = EnumUtils.fromValue(Color::class.java, "red")
    assertEquals(Color.RED, result)
  }

  @Test
  fun `fromValue throws IllegalArgumentException for unknown value`() {
    val exception =
        assertThrows(IllegalArgumentException::class.java) {
          EnumUtils.fromValue(Color::class.java, "yellow")
        }
    assertEquals("Unknown Color value: yellow", exception.message)
  }

  @Test
  fun `fromValue exception message includes the enum class name`() {
    val exception =
        assertThrows(IllegalArgumentException::class.java) {
          EnumUtils.fromValue(Priority::class.java, 99)
        }
    assertEquals("Unknown Priority value: 99", exception.message)
  }
}
